File: src/main/java/org/hwabeag/hwaskyblock/events/BreakEvent.java
package org.hwabeag.hwaskyblock.events;

import org.bukkit.ChatColor;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.hwabeag.hwaskyblock.config.ConfigManager;

import java.util.Objects;

public class BreakEvent implements Listener {

    FileConfiguration Config = ConfigManager.getConfig("setting");
    FileConfiguration MessageConfig = ConfigManager.getConfig("message");
    FileConfiguration SkyBlockConfig = ConfigManager.getConfig("skyblock");
    String Prefix = ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("hwaskyblock-system.prefix")));

    @EventHandler
    public void onBreak(BlockBreakEvent event) {
        Player player = event.getPlayer();
        String name = player.getName();
        Block block = event.getBlock();
        World world = block.getWorld();
        String world_name = world.getWorldFolder().getName();
        String[] number = world_name.split("\\.");
        if (Objects.equals(number[0], "HwaSkyBlock")) {
            String id = number[1];
            if (SkyBlockConfig.getString(id + ".leader") != null) {
                if (!Objects.equals(SkyBlockConfig.getString(id + ".leader"), name)) {
                    if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                        if (!SkyBlockConfig.getBoolean(id + ".break")) {
                            player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(MessageConfig.getString("message-event.no_permission"))));
                            event.setCancelled(true);
                            return;
                        }
                    } else {
                        if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".break")) {
                            player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(MessageConfig.getString("message-event.no_permission"))));
                            event.setCancelled(true);
                            return;
                        }
                    }
                    if (!isInRegion(event.getBlock(), id)) {
                        event.setCancelled(true);
                    }
                }
            }
        }
    }

    private boolean isInRegion(Block block, String id) {
        int x = block.getX();
        int y = block.getY();
        int z = block.getZ();
        return (x >= 0 && x < SkyBlockConfig.getInt(id + ".size")) &&
                (y >= 0 && y < 256) &&
                (z >= 0 && z < SkyBlockConfig.getInt(id + ".size"));
    }
}

File: src/main/kotlin/events/block/BreakEvent.kt
package org.hwabeag.hwaskyblock.events.block

import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.ChatColor
import org.bukkit.block.Block
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.hwabeag.hwaskyblock.database.DatabaseManager
import org.hwabeag.hwaskyblock.database.config.ConfigManager
import java.util.*

class BreakEvent : Listener {
    var Config: FileConfiguration = ConfigManager.getConfig("setting")!!
    var MessageConfig: FileConfiguration = ConfigManager.getConfig("message")!!
    var Prefix: String = ChatColor.translateAlternateColorCodes(
        '&',
        Objects.requireNonNull<String?>(Config.getString("hwaskyblock-system.prefix"))
    )

    @EventHandler
    fun onBreak(event: BlockBreakEvent) {
        val player = event.player
        val name = player.name
        val block = event.getBlock()
        val world = block.world
        val world_name = world.worldFolder.name
        val number: Array<String?> = world_name.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        if (number[0] == "HwaSkyBlock") {
            val id = number[1]
            if (DatabaseManager.getSkyBlockData(id.toString(), "$id.leader", "getSkyBlockLeader") != null) {
                val leader =
                    DatabaseManager.getSkyBlockData(id.toString(), "$id.leader", "getSkyBlockLeader") as? String
                if (leader != name) {
                    val isSharer = DatabaseManager.getSkyBlockShareList(id.toString()).contains(name)

                    val hasBreakPermission = if (!isSharer) {
                        DatabaseManager.getSkyBlockData(id.toString(), "$id.break", "isSkyBlockBreak") as? Boolean
                            ?: false
                    } else {
                        DatabaseManager.getSkyBlockSharePermission(id.toString(), name, "break") ?: false
                    }

                    if (!hasBreakPermission) {
                        val message = ChatColor.translateAlternateColorCodes(
                            '&',
                            Prefix + (MessageConfig.getString("message-event.no_permission") ?: "")
                        )
                        player.spigot().sendMessage(ChatMessageType.ACTION_BAR, TextComponent(message))
                        event.isCancelled = true
                        return
                    }

                    if (!isInRegion(event.block, id)) {
                        event.isCancelled = true
                    }
                }
            }
        }
    }


    private fun isInRegion(block: Block, id: String?): Boolean {
        val x = block.x
        val y = block.y
        val z = block.z
        val size = (DatabaseManager.getSkyBlockData(id.toString(), "$id.size", "getSkyBlockSize") as? Int) ?: 0
        return (x >= 0 && x < size) &&
                (y >= 0 && y < 256) &&
                (z >= 0 && z < size)

    }
}